pkg/bundle/sync.go

```diff
@@ -259,137 +262,265 @@ func (e pkcs12Encoder) encode(trustBundle string) ([]byte, error) {
 // The name of the ConfigMap is the same as the Bundle.
 // Ensures the ConfigMap is owned by the given Bundle, and the data is up to date.
 // Returns true if the ConfigMap has been created or was updated.
-func (b *bundle) syncTarget(ctx context.Context, log logr.Logger,
+func (b *bundle) syncTarget(
+	ctx context.Context,
+	log logr.Logger,
 	bundle *trustapi.Bundle,
-	namespaceSelector labels.Selector,
-	namespace *corev1.Namespace,
+	name string,
+	namespace string,
 	data string,
+	shouldExist bool,
 ) (bool, error) {
-	target := bundle.Spec.Target
+	configMap := &metav1.PartialObjectMetadata{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+	}
+	err := b.targetCache.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, configMap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("failed to get ConfigMap %s/%s: %w", namespace, name, err)
+	}
 
+	// If the ConfigMap exists, but the Bundle is being deleted, delete the ConfigMap.
+	if apierrors.IsNotFound(err) && !shouldExist {
+		return false, nil
+	}
+
+	// If the ConfigMap should not exist, but it does, delete it.
+	if !apierrors.IsNotFound(err) && !shouldExist {
+		// apply empty patch to remove the key
+		configMapPatch := coreapplyconfig.
+			ConfigMap(name, namespace).
+			WithLabels(map[string]string{
+				trustapi.BundleLabelKey: bundle.Name,
+			}).
+			WithOwnerReferences(
+				v1.OwnerReference().
+					WithAPIVersion(trustapi.SchemeGroupVersion.String()).
+					WithKind(trustapi.BundleKind).
+					WithName(bundle.GetName()).
+					WithUID(bundle.GetUID()).
+					WithBlockOwnerDeletion(true).
+					WithController(true),
+			)
+
+		if err = b.patchResource(ctx, configMapPatch); err != nil {
+			return false, fmt.Errorf("failed to patch configMap %s/%s: %w", namespace, bundle.Name, err)
+		}
+
+		return true, nil
+	}
+
+	target := bundle.Spec.Target
 	if target.ConfigMap == nil {
 		return false, errors.New("target not defined")
 	}
 
-	matchNamespace := namespaceSelector.Matches(labels.Set(namespace.Labels))
-
-	var configMap corev1.ConfigMap
-	err := b.client.Get(ctx, client.ObjectKey{Namespace: namespace.Name, Name: bundle.Name}, &configMap)
+	dataHash := fmt.Sprintf("%x", sha256.Sum256([]byte(data)))
```

Comment:
You have maybe considered this already, but working on the PKCS12 feature, I noted this comment - which is deleted in this PR:

> Generated JKS is not deterministic - best we can do here is update if the pem cert has changed (hence not checking if JKS matches)

If this holds, I think we might have a potential reconcile loop form the logic using a hash.

Reply:
The hash should only hash the data PEM string.
We only use the hash to determine if the "source of truth" PEM bundle changed.
Any modification to existing PEM contents/ JKS contents/ ... we notice by making sure we are still field owner of the key in the map.

Reply:
Right, `data` contains the PEM only. The `binaryData` field will container JKS/PKCS12. My bad.

Reply:
```suggestion
        // Generated JKS is not deterministic - best we can do here is update if the pem cert has
	// changed (hence not checking if JKS matches)
	dataHash := fmt.Sprintf("%x", sha256.Sum256([]byte(data)))
```

I think it's worth keeping this comment for future reference. WDYT?